Src/Particle/AMReX_DenseBins.H

```diff
@@ -0,0 +1,117 @@
+#ifndef AMREX_DENSEBINS_H_
+#define AMREX_DENSEBINS_H_
+
+#include <AMReX_Gpu.H>
+#include <AMReX_IntVect.H>
+
+namespace amrex
+{
+
+/**
+ * \brief A container for storing items in a set of bins.
+ * 
+ * The underlying data structure is an array of size nitems defining a
+ * permutation of the items in the container that puts them in bin-sorted order,
+ * plus an array of size nbins that stores the offsets into the permutation
+ * array where each bin starts.
+ *
+ * The storage for the bins is "dense" in the sense that users pass in
+ * a Box that defines the space over which the bins will be defined, and
+ * empty bins will still take up space.
+ *
+ * \tparam The type of items we hold
+ *
+ */    
+template <typename T>
+class DenseBins
+{
+    using bin_type = IntVect;
+    using index_type = unsigned int;
+    
+public:
+
+    /**
+     * \brief Populate the bins with a set of items.
+     *
+     * The algorithm is similar to a counting sort. First, we count the number
+     * of items in each bin. Then, we perform a prefix sum on the resulting counts.
+     * Finally, the set of partial sums is incremented in parallel using atomicInc,
+     * which results in a permutationa array that places the items in bin-sorted order.
+     * 
+     * \tparam N the 'size' type that can enumerate all the items
+     * \tparam F a function that maps items to IntVect bins
+     *
+     * \param nitems the number of items to put in the bins
+     * \param v pointer to the start of the items
+     * \param bx the Box that defines the space over which the bins will be defined
+     * \param f a function object that maps items to bins 
+     */        
+    template <typename N, typename F>
+    void build (N nitems, T const* v, const Box& bx, F f)
+    {
+        BL_PROFILE("DenseBins<T>::build");
+
+        m_cells.resize(nitems);
+        m_perm.resize(nitems);
+        
+        auto nbins = bx.numPts();
+        m_counts.resize(0);
+        m_counts.resize(nbins, 0);
+
+        m_offsets.resize(0);
+        m_offsets.resize(nbins+1);
+
+        const auto lo = lbound(bx);
+        const auto hi = ubound(bx);
+        index_type* pcell   = m_cells.dataPtr();
+        index_type* pcount  = m_counts.dataPtr();
+        index_type* poffset = m_offsets.dataPtr();
+        AMREX_FOR_1D ( nitems, i,
+        {
+            bin_type iv = f(v[i]);
+            int nx = hi.x-lo.x+1;
+            int ny = hi.y-lo.y+1;
+            int nz = hi.z-lo.z+1;
+            index_type uix = min(nx-1,max(0,iv[0]));
+            index_type uiy = min(ny-1,max(0,iv[1]));
+            index_type uiz = min(nz-1,max(0,iv[2]));
+            pcell[i] = (uix * ny + uiy) * nz + uiz;
+            Gpu::Atomic::Add(&pcount[pcell[i]], index_type{ 1 });
+        });
+
+        Gpu::exclusive_scan(m_counts.begin(), m_counts.end()+1, m_offsets.begin());
```

Comment:
Why `end()+1`?

Reply:
offsets is size num_bins + 1, so that the last entry is the total number of items. Since this is an exclusive scan, the last value in the input array is never used. But this way, the last entry of  the output will be written to.

Reply:
Note, this is now implemented by making `m_counts` size `nbins+1 as well,